### PR TITLE
[PyTorch] Fix stale columnwise data usage

### PR DESCRIPTION
**Pull request by ksivaman**
# Description

This PR sets columnwise usage correctly for all quantizers instead of retaining the value in the quantizer, which may be incorrect after resuming training post validation steps as the columnwise usage is set to `False` for eval mode.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Set columnwise usage explicitly instead of retaining the value in the quantizer.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/test_sanity.py [hunk 1]
```diff
@@ -1233,3 +1233,92 @@ def check_weights():
...
+    run_forward(is_eval=False)
+    for q in get_weight_quantizers():
+        assert (
+            q.columnwise_usage
+        ), "After an initial training forward, weight quantizer should have columnwise_usage=True"
+
+    # 2. Eval forward -- should set columnwise=False on primary FP8 weight
+    # quantizers, simulating the start of an evaluation loop.
+    run_forward(is_eval=True)
+    for q in get_weight_quantizers():
```
Comment (vthumbe1503) [inline tests/pytorch/test_sanity.py:1313]: The test doesnt make sense to me. I dont think we should be toggling the quantizer usages in case of quantized_model_init at all. 

This breaks the very principle that quantized tensor and its internal quantizer shouldnt be in conflict with each other.  And in here, the columnwise_data is present for the quantized_tensor even though the columnwise_usage is set to False.

File: transformer_engine/pytorch/module/grouped_linear.py [hunk 2]
```diff
@@ -135,16 +135,11 @@ def forward(
...
                 )
-            # No need to set the quantizer states if weight is already quantized
-            # for debug mode we create quantizer every iteration, thus we need to set the quantizer states
-            if weight_quantizers[0] is not None and (
-                not isinstance(weights[0], QuantizedTensorStorage) or debug
-            ):
+            if weight_quantizers[0] is not None:
+                if isinstance(weights[0], QuantizedTensorStorage) and not debug:
+                    weight_quantizers = [weight._quantizer for weight in weights]
                 for weight_quantizer in weight_quantizers:
```
Comment (vthumbe1503) [inline transformer_engine/pytorch/module/grouped_linear.py:141]: I dont think these changes are needed for this file and any of the other files in the PR. If weight is already quantized, it doesnt make sense to change its internal quantizer and have the quantized weight and its internal quantizer in a state of conflict with each other.

In general in case of quantized_model_init, if we are changing quantized_tensor's internal quantizer, quantized_tensor should also be updated to have that appropriate usages.

File: transformer_engine/pytorch/tensor/float8_blockwise_tensor.py [hunk 3]
```diff
@@ -653,14 +653,14 @@ def fsdp_pre_all_gather(self, mesh, orig_size, contiguous_orig_stride, module, m
...
+        is_backward_pass = training_state == TrainingState.PRE_BACKWARD
         if reshard_after_forward:
-            training_state = param_group._training_state
-            is_backward_pass = training_state == TrainingState.PRE_BACKWARD
             rowwise_usage = not is_backward_pass
             columnwise_usage = is_backward_pass
         else:
             rowwise_usage = True
-            columnwise_usage = self._quantizer.columnwise_usage
+            columnwise_usage = is_backward_pass or torch.is_grad_enabled()
```
Comment (vthumbe1503) [inline transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:663]: I think it still makes sense to use self._quantize.columwise_usage as the real truth of what data is 
"really avaliable" in the sharded quantized tensor and throw an error if that usage doesnt match
` is_backward_pass or torch.is_grad_enabled `(Similar to mxfp8 tensor)

What we are doing here is that we are silently creating columnwise data after allgather for allgathered tensor, even though original sharded data tensor didnt have that data.

In my opinion, I am against any change here since even doing such a validation and throwing error is going to incur CPU overheads when using 
`torch.is_grad_enabled `

Comment (vthumbe1503) [reply transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:663]: Same comment in every other FSDP2 related changes
